src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -59,6 +61,72 @@ to_soa_layout(const std::vector<ingress_egress_state<T>>& v) {
 
 namespace {
 
+const auto direction_label = ss::metrics::label("direction");
+const auto label_ingress = direction_label("ingress");
+const auto label_egress = direction_label("egress");
+
+template<class T>
+void add_ingress_egress_gauges(
```

Comment:
You probably mean this:
```
template<class T>
void add_ingress_egress_gauges(...,
  std::function<ingress_egress_state<T>())&& value_getter, ...)
```

If so, the problem with this code is that the compiler cannot automatically deduce `T` from that context.

Reply:
No , the above actually works. This compiled and ran for me:
```
#include <iostream>

template<typename T>
struct MyStruct{
    T value;
};

template<typename T>
void printValue(MyStruct<T> t)
{
    std::cout << "t: " << t.value;
}

int main() {
    MyStruct<int> t;
    t.value = 15;
    printValue(t);
    return 0;
}
```

Reply:
That works for sure, it's just a deduction from a class template paramter. In my case, I would need the compiler to deduce `T` from a parameter of a class template that is a return type of a function parameter for my template. 
```
#include <iostream>

template<typename T>
struct MyStruct{
    T value;
};

template<typename T>
void printValue(std::function<MyStruct<T>()> f)
{
    std::cout << "t: " << f().value;
}

int main() {
    printValue([](){
        MyStruct<int> t;
        t.value = 15;
        return t;
    });
    return 0;
}
```

(godbolt)[https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjE/lykAA6oCoRODB7evnrJqY4CIWGRLDFxCXaYDulCBEzEBJk%2BfgEVVQI1dQSFEdGx8ba19Y3Z5YNdoT0lfVwAlLaoXsTI7BzmAMyhyN5YANQma25OCgTEmKz72CYaAIJX1wSYLIkGD/tuBACeiYysmDsAKhc7sdiF4HDsALIfGqghwmADsVhuOxRAJ2ADcxF5MPskbd4QARXF3O4PJ4vHEHT7fZhsAFAm7o1B4dA7RLEUIEABqWMwEGO6BAICoXgYbQYbyhMLBBDegLW2AgMwuOyoypuCLxqJ2AqFaC8BD2BzeezMZgIIFNZiNbhNVCVADpMd5KXiEUSNTdOTsWExQkq9oi7tr2ZyeS6ICYAKxWKMEpWa4Pa1FSk4yt6clWytZa5MoghO3lGgk7LhR4nIvM7U4ERYMHbZ3Mo93KnNJlE1us7DQV/EEjhzWicKO8PwcLSkVCcW2Waw6hZLP7rHikC3jgdzADWIEkawdAA591H92s1gBOffws9RyRn%2BH6TiSUeaXhTji8BQgDSrl9zOCwGBEBQVAnjoWJyEoNBQPoOItkMYAyw0b8sHRPBli5PBMAAdwAeRpTgVxoWgHmIT8ICiF9SCiUI6g%2BAjeGo5hiA%2BHCom0So1xXKC2EEHCGFoOj11ILBfSMcQhPwU4qnRTBPyEzBVEqA0VhXTlMCHITaDwKJiFojwsEok48BYei5ioAxgAUDDsLwxh6JkQQRDEdgpAc%2BQlDUSjdASAwjBQaxrH0bTP0gOZUESfIGDknYAFoBX2AlTFnSwzA0WKBS0qIEsypKrEsWKcInVpIpcBh3E8Jo9GCCZilKHIUjSARhj8BJckahhulq6ZbHUjjqjGZq9GK/rOk63oygGTpBtGUaavGiQ5gUBdlgWh8OBHUgxwnN8dlUfcADYYv2yQdjgoxSyjB0NCunYIFwQgSFNNZZl4NctBmLcQFPB1T1%2Bv7/v2tan1IEy1n2h1jovDQ1jMfaNDMY8ozPTbKLfD8vx/dc/0AiAkAWAhEgNCCICgxIwOIcJfk4OoWHRfcYqYU7fOAUsuCuh0V0wfAiA5QUEn4RzRHELgAgF9yVHUITvNILDdMSUy1o2rbX04HCDUJw1UCoXaDqOk6zpZst2bSiAPGg2InpezH3rmBAziwOIlSB3gTLMSQfskfcuC4E8uDPM8kP2gJlcnTh0e/N6B1If8gIUpSeeJupLOUQx1KEBBUCwscuJAsmXnSFOwlodPM5D0nyZQZnEO/cuYLV4QM6zyi4%2BQa5iEssPAkU5AanwMdeDFpzhekMXFAlry9GZ/zkpsTKQqd8LIuiuKCHQBLcusVL0tXzLsu0jf8piwqP2WlyBj7wu08b7PXtOFTeFlph5e4KONKV1HOGwbuDUevbDuOpm8ELrG1ujOPKNgdj3R5pbGYr1fyDkfC7L6Z4rrww0PCfckg/aYI0F7aQIc0a2AxpHD6pBtxRnBt7LB%2B04Yiw0JIKM0M1prGfEJQhJCEEcDMKw7ancOGkBkqRdIO4gA]

I guess this is not possible because functions are not overloded by return types.

Reply:
It works if you don't use `std::function` [godbolt](https://godbolt.org/z/oezGzEb9a). Looks like `make_gauge` doesn't use `std::function` either so it should work.

Reply:
> Looks like make_gauge doesn't use std::function either

It doesn't use `std::function` explicitly, but it accepts a callable as well as a value there.

```
template<typename T>
impl::metric_definition_impl make_gauge(metric_name_type name,
        T&& val, description d=description(), std::vector<label_instance> labels = {}) {
    return {name, {impl::data_type::GAUGE, "gauge"}, make_function(std::forward<T>(val), impl::data_type::GAUGE), d, labels};
}

template<typename T, typename = std::enable_if_t<is_callable<T>::value>>
metric_function make_function(T val, data_type dt) {
    return [dt, val] {
        return metric_value(val(), dt);
    };
}

template<typename T, typename = std::enable_if_t<!is_callable<T>::value>>
metric_function make_function(T& val, data_type dt) {
    return [dt, &val] {
        return metric_value(val, dt);
    };
}
```